File: python/data_structures/queue.py
from data_structures.node import Node


class Queue:
    head = None
    tail = None

    def __init__(self, lst=None):
        lst = lst or []

        for value in lst:
            self.push(value)

    def push(self, value):
        new_node = Node(value)

        if self.head is None:
            self.head = new_node
            self.tail = self.head
        else:
            self.tail.next = new_node
            self.tail = new_node

    def pop(self):
        if self.head:
            data = self.head.data
            self.head = self.head.next
            return data

    def peek(self):
        return self.head.data

File: golang/data_structures/queue/queue.go
package stack

import (
	"fmt"
	"strconv"
	"strings"
)

type Node struct {
	value int
	next  *Node
}

func NewNode(value int) *Node {
	node := new(Node)
	node.value = value

	return node
}

type Queue struct {
	head *Node
	tail *Node
}

func NewQueue(args ...int) *Queue {
	list := new(Queue)

	for _, v := range args {
		list.enqueue(v)
	}

	return list
}

func (q *Queue) String() string {
	node := q.head

	var values []string

	if node == nil {
		return "[]"
	}

	for node.next != nil {
		values = append(values, strconv.Itoa(node.value))

		node = node.next
	}
	values = append(values, strconv.Itoa(node.value))

	return fmt.Sprintf("[%s]", strings.Join(values, ", "))
}

func (q *Queue) enqueue(value int) {
	newTail := NewNode(value)

	if q.head == nil {
		q.head = newTail
		q.tail = q.head
	} else {
		q.tail.next = newTail
		q.tail = newTail
	}
}

func (q *Queue) dequeue() (value int, ok bool) {

	if q.head == nil {
		return 0, false
	} else {
		node := q.head
		q.head = node.next
		return node.value, true
	}
}

func (q *Queue) peek() (value int, ok bool) {
	if q.head == nil {
		return 0, false
	} else {
		return q.head.value, true
	}
}
